File: Python/upgma_lib.py
import numpy as np
import math


class ElementaryNode():
    """
    Class describing a leaf in a Newick tree, only contains the alphabetic name
    and the attribute minimum which is set to zero and exists for the sole purpose
    of being compatible with the NewickNode class constructor.
    """
    def __init__(self, name):
        self.name = name
        self.minimum = 0.


class NewickNode():
    """
    Class describing a node in a Newick tree, constructed from it's children
    children can be of same type or of ElementaryNode type
    """
    def __init__(self, children, minimum):
        self.children = children
        self.minimum = minimum/2.
        output = "("
        output += self.children[0].name
        output += ":"+str(self.minimum-self.children[0].minimum)
        output += ","
        output += self.children[1].name
        output += ":"+str(self.minimum-self.children[1].minimum)
        output += ")"
        #output += ":"+str(self.minimum/2.)
        self.name = output


def createNameList():
    """
    Creates a list with potential alphabetic names, starting with "A",
    ending with "ZZ", containing all possible variations.

    Input:
    none

    Output:
    nameList: python list object
    """
    nameList = []
    for i in range(ord("Z")+1-ord("A")):
        nameList.append(chr(ord("A")+i))
    for i in range(ord("Z")+1-ord("A")):
        for j in range(ord("Z")+1-ord("A")):
            nameList.append(chr(ord("A")+i)+chr(ord("A")+j))

    return nameList


def findMinimumValue(matrix):
    """
    Finds the minimum value in the lower triangulate of a square matrix,
    complexity O(nlog(n))

    Input:
    matrix: numpy matrix object

    Output:
    python list object containing the minimum value and its indices
    """
    minValue = matrix[0][1]
    minI = 0
    minJ = 1
    for i in range(len(matrix)):
        for j in range(i):
            if matrix[i][j] != 0 and matrix[i][j] < minValue:
                minValue = matrix[i][j]
                minI = i
                minJ = j

    return [minValue, minI, minJ]


def readMatrixfromFile(filename):
    """
    Reads from an input file and creates a numpy matrix object.

    Input:
    file: python file object

    Output:
    matrix: numpy matrix object
    """
    return np.loadtxt(filename)


def jukesCantor(fileObject, typeOfInput):
    """
    Reads from an input file containing aligned sequences and creates a dissimilarity matrix.
    Uses the Jukes Cantor (JC69) DNA Model
    Input:
    file: python file object

    Output:
    matrix: numpy matrix object
    """
    sequences = []
    nameList = []
    tmpSequence = ""
    if typeOfInput == "FASTA":
        for line in fileObject:
            if line[0] == '>':
                nameList.append(line.strip(">"))
                if tmpSequence != "":
                    sequences.append(tmpSequence)
                    tmpSequence = ""
            else:
                tmpSequence += line.strip()
        sequences.append(tmpSequence)

    else:
        for line in fileObject:
            sequences.append(line.strip())


    sequenceLength = len(sequences[0])
    numberOfSequences = len(sequences)

    matrix = np.zeros((numberOfSequences, numberOfSequences))

    #print numberOfSequences

    for i in range(numberOfSequences-1):
        for j in range(i+1,numberOfSequences):
            numdiff = 0
            for index in range(sequenceLength):
                if sequences[i][index] != sequences[j][index]:
                    numdiff += 1

            if (float(numdiff)/float(sequenceLength)) >= 0.75:
                return "Error"
            distance = -0.75 * math.log(1 - (4./3.)*(float(numdiff) / sequenceLength), math.e)
            matrix[i][j] = distance
            matrix[j][i] = distance
        print i

    return matrix, nameList

def kimura(fileObject, typeOfInput):
    """
    Reads from an input file containing aligned sequences and creates a dissimilarity matrix.
    Uses the Kimura (K80) DNA Model

    Input:
    file: python file object

    Output:
    matrix: numpy matrix object
    """
    sequences = []
    nameList = []
    tmpSequence = ""

    if typeOfInput == "FASTA":
        for line in fileObject:
            if line[0] == '>':
                nameList.append(line.strip(">"))
                if tmpSequence != "":
                    sequences.append(tmpSequence)
                    tmpSequence = ""
            else:
                tmpSequence += line.strip()
        sequences.append(tmpSequence)

    else:
        for line in fileObject:
            sequences.append(line.strip())


    sequenceLength = len(sequences[0])
    numberOfSequences = len(sequences)

    matrix = np.zeros((numberOfSequences, numberOfSequences))

    for i in range(numberOfSequences-1):
        for j in range(i+1, numberOfSequences):
            numTransitions = 0.
            numTransversions = 0.
            for index in range(sequenceLength):
                case = ((sequences[i][index] == 'A' and sequences[j][index] == 'G') or
                        (sequences[i][index] == 'G' and sequences[j][index] == 'A') or
                        (sequences[i][index] == 'T' and sequences[j][index] == 'C') or
                        (sequences[i][index] == 'C' and sequences[j][index] == 'T'))
                if case:
                    numTransitions += 1
                elif sequences[i][index] == "-" or sequences[j][index] == "-" or sequences[i][index]==sequences[j][index]:
                    pass
                else:
                    numTransversions += 1

            numTransversions /= sequenceLength
            numTransitions /= sequenceLength
            distance = -0.5*math.log(1-numTransversions-2*numTransitions, math.e) - 0.25*math.log(1-2*numTransversions,
                                                                                                  math.e)
            matrix[i][j] = distance
            matrix[j][i] = distance
        print i

    return matrix, nameList

File: Ruby/upgma_lib.rb
require 'narray'

def pass

end

class ElementaryNode
  def initialize(name)
    @name = name
    @minimum = 0
  end
end

  class NewickNode
    def initialize(children, minimum)
      @children = children
      @minimum = minimum.to_f / 2
      output = "("
      output += @children[0].instance_variable_get(:@name)
      output += ":"+(@minimum-@children[0].instance_variable_get(:@minimum)).to_s
      output += ","
      output += @children[1].instance_variable_get(:@name)
      output += ":"+(@minimum-@children[1].instance_variable_get(:@minimum)).to_s
      output += ")"
      @name = output
    end
  end

  def createNamingList()
    list = []
    for i in 0...("Z".ord + 1 - "A".ord)
      list.push(("A".ord+i).chr)
    end
    for i in 0...("Z".ord + 1 - "A".ord)
      for j in 0...("Z".ord + 1 - "A".ord)
        list.push(("A".ord+i).chr + ("A".ord+j).chr)
      end
    end
    list
  end

  def findMinimumValue(matrix)

    #   Finds the minimum value in the lower triangulate of a square matrix,
    #   complexity O(nlog(n))
    #
    #   Input:
    #   matrix: narray matrix object
    #
    #   Output:
    #   ruby list object containing the minimum value and its indice

    min = matrix[0, 1]
    minI = 0
    minJ = 1
    for i in 0...(matrix.shape[0])
      for j in 0..(i)
        if (matrix[i, j] != 0 and matrix[i, j] < min)
          min = matrix[i, j]
          minI = i
          minJ = j
        end
      end
    end
    [min, minI, minJ]
  end

  def to_readable(matrix)

    # Prints matrix in more readable style

    i = 0
    matrix.each do |number|
      print number.to_s + " "
      i+= 1
      if i == matrix.shape[0]
        print "\n"
        i = 0
      end
    end
  end

  def jukesCantor(file, type)

    #   Jukes-Cantor algorithm
    #   Reads from an input file containing aligned sequences and creates a dissimilarity matrix.
    #
    #   Input:
    #   file: ruby file object
    #
    #   Output:
    #   matrix: narray matrix object

    sequences = []
    tmpSequence = ""
    if type == "FASTA"
      for line in file
        if line[0] == '>'
          if tmpSequence != ""
            sequences.push(tmpSequence)
            tmpSequence = ""
          end
        else
          tmpSequence += line.strip
        end
      end
      sequences.push(tmpSequence)
    else
      for line in file
        sequences.push(line.strip)
      end
    end

    matrix = NArray.float(sequences.length, sequences.length)

    i = 0
    j = 0

    sequences.each { |seq1|
      j = 0
      sequences.each { |seq2|
        if i > j
          j += 1
          next
        elsif i == j
          matrix[i, j] = 0
        else
          numdiff = 0
          for index in 0...(seq1.length)
            if seq1[index] != seq2[index]
              numdiff += 1
            end
          end
          if (numdiff == seq1.length) or ((numdiff.to_f/seq1.length) >= 0.75)
            return "Error"
          end
          numdiff = numdiff.to_f / seq1.length
          distance = -0.75 * Math.log((1 - (4.to_f / 3) * numdiff), Math::E)
          matrix[i, j] = distance
          matrix[j, i] = distance
        end
        j += 1
      }
      i += 1
#      print i
#      puts
    }
    matrix
  end

def kimura(file, type)

  #   Kimura algorithm
  #   Reads from an input file containing aligned sequences and creates a dissimilarity matrix.
  #
  #   Input:
  #   file: ruby file object
  #
  #   Output:
  #   matrix: narray matrix object

  sequences = []
  tmpSequence = ""
  if type == "FASTA"
    for line in file
      if line[0] == '>'
        if tmpSequence != ""
          sequences.push(tmpSequence)
          tmpSequence = ""
        end
      else
        tmpSequence += line.strip
      end
    end
    sequences.push(tmpSequence)
  else
    for line in file
      sequences.push(line.strip)
    end
  end

  matrix = NArray.float(sequences.length, sequences.length)

  i = 0
  j = 0

  sequences.each { |seq1|
    j = 0
    sequences.each { |seq2|
      if i > j
        j += 1
        next
      elsif i == j
        matrix[i, j] = 0
      else
        translation = 0
        transversion = 0
        for index in 0...(seq1.length)
          if seq1[index] != seq2[index]
            if (seq1[index] == 'A' and seq2[index] == 'G') or (seq1[index] == 'G' and seq2[index] == 'A') or (seq1[index] == 'T' and seq2[index] == 'C') or (seq1[index] == 'C' and seq2[index] == 'T')
              translation += 1
            elsif seq1[index] == '-' or seq2[index] == '-'
              pass
            else
              transversion += 1
            end
          end
        end
        translation = translation.to_f / seq1.length
        transversion = transversion.to_f / seq1.length
        distance = -0.5 * Math.log(1 - 2 * translation - transversion, Math::E) -0.25 * Math.log(1 - 2 * transversion, Math::E)
        matrix[i, j] = distance
        matrix[j, i] = distance
      end
      j += 1
    }
    i += 1
#    print i
#    puts
  }
  matrix
end